### ISSUE TITLE
Add upload progress callback

### ISSUE THREAD

**Issue by psixdev** [sindresorhus/ky#183]
Hello, I would like to add an image loading indicator. Is this possible when using this library?

**Comment by sholladay** [comment]
Did you try the `onDownloadProgress` callback?

**Comment by psixdev** [comment]
Sorry, it seems I did not ask the question correctly. I need upload progress, not download progress. And onDownloadProgress doesn't help me.

**Comment by sholladay** [comment]
I see. We don't have that feature at the moment, but it would be nice to add. I think it could be done similarly to download progress.

**Comment by Qix-** [comment]
It cannot. `ky` uses `fetch` under the hood, which uses streams, which does not support upload progress callbacks.

Seems like a pretty bizarre thing to leave out of the spec, if you ask me. Until they're added, `ky` cannot support them unless it falls back to `XMLHttpRequest`.

**Comment by sholladay** [comment]
I realize that `fetch` doesn't have a callback for upload progress, which is a shame. But I think we should be able to detect when `options.body` is a stream and then send it through a passthrough stream of our own that reads the chunks manually, similarly to how our `onDownloadProgress` option works.

**Comment by Qix-** [comment]
As long as we can determine the total bytes to be uploaded then absolutely :)

**Comment by sholladay** [comment]
I suppose if the body is just a plain stream, we'd only be able to provide how many bytes have been uploaded but not the total bytes, which may not be super useful.

However, if the body is a [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob) or a [File](https://developer.mozilla.org/en-US/docs/Web/API/File) (or anything that can be normalized to one of those), then we can use `blob.size` to get the total bytes. And then we can use `blob.stream()` to get a `ReadableStream` for it.

**Comment by aldipower** [comment]
Really missing this for File uploads. You possible solution sound great @sholladay .

**Comment by supertiger1234** [comment]
can this be added yet? axios seems to have this feature.

**Comment by Qix-** [comment]
@supertiger1234 You are perfectly allowed to submit a PR, yes :)

**Comment by supertiger1234** [comment]
🤷 I wouldn't even be using a lib for fetching if I knew how.

**Comment by IRediTOTO** [comment]
Please make this :D

**Comment by sholladay** [comment]
For anyone who wants to try implementing this, take a look at https://github.com/sindresorhus/ky/pull/34 and use that as a reference. Once there's a rough draft, I'd be happy to review and help with the code.

**Comment by Murderlon** [comment]
For anyone who wants to have it work right now, I tested the following with regular `File` uploads and `FormData` uploads and it works:

```js
function onUploadProgress(stream, callback) {
  const reader = stream.getReader()
  let bytesUploaded = 0

  return new ReadableStream({
    start(controller) {
      async function push() {
        try {
          const { done, value } = await reader.read()

          if (done) {
            controller.close()
            return
          }

          bytesUploaded += value.length
          callback(bytesUploaded)
          controller.enqueue(value)
          push()
        } catch (error) {
          controller.error(error)
        }
      }
      push()
    },
  })
}

onUploadProgress(file.stream(), (bytesUploaded) => console.log(bytesUploaded))

await ky('some-endpoint', { body: file })
```

**Comment by Murderlon** [comment]
> It seems this should now be possible with Chromium versions > 105. It would be great if it could be incorporated into ky!

I've looked into this a bit more and it wouldn't be as easy to implement in `ky`. The presence of `onUploadProgress` would have to force the passing of `duplex: 'half'` to  underlying `fetch` and force the `body` to be a stream. If you already have the entire file, you have to manually push it through a stream again. When uploading `FormData`, this would result in hacky conversions which I haven't gotten to work at all. Further more, supporting `onUploadProgress` would mean [dropping support for HTTP/1.1](https://developer.chrome.com/articles/fetch-streaming-requests/#doesnt-work-on-http1x), which frankly is unacceptable. Even giants like AWS S3 still only support uploading over HTTP/1.1. 

All in all, I don't see an addition like this making sense until `fetch` itself exposes more control over this (such as some sort of [fetch observer](https://github.com/whatwg/fetch/issues/607))

**Comment by jadedevin13** [comment]
@Murderlon  https://github.com/sindresorhus/ky/pull/632 would you mind giving a review?

**Comment by mifi** [comment]
The problem with #632 is that AFAIK it will crash in browsers other than Chrome which is not really acceptable IMO.

FWIW [this person on x](https://x.com/vvoyer/status/1856287377188368449) implemented a tiny wrapper which uses fetch if the client supports upload progress over fetch, with fallback to XMLHTTPRequest:
- https://github.com/vercel/storage/blob/main/packages/blob/src/request.ts
- https://github.com/vercel/storage/blob/main/packages/blob/src/xhr.ts
- https://github.com/vercel/storage/blob/main/packages/blob/src/fetch.ts
- https://github.com/vercel/storage/blob/3491560f22d54eebaac8310a5b3df38fafbb58f3/packages/blob/src/helpers.ts#L131

I'm not sure if `ky` could do something like that, e.g. whether ky's API surface can be satisfied using an XMLHttpRequest backend (in addition to `fetch`).

Alternatively one could make a modified version of the [fetch polyfill](https://github.com/JakeChampion/fetch) that exposes its inner XMLHttpRequest object, then ky could hook into its `progress` event which could be used as a fallback in #632

**Comment by sholladay** [comment]
> The problem with #632 is that AFAIK it will crash in browsers other than Chrome which is not really acceptable IMO.

True, but only if you provide an `onUploadProgress` function, which is a new option. It has no impact on existing apps. This is consistent with our approach to other features that don't have wide browser support. Obviously if you need to support Firefox, then you can't use this feature yet. It's still helpful for Chrome users.

The only real alternative here is for us to ignore the option instead of throwing when it's not supported. But silently failing is bad behavior for a library. Instead, you could detect support in your app and selectively provide the `onUploadProgress` option to achieve the same effect.

To that end, I would be supportive of Ky exporting its feature detection mechanisms. Or moving them to [`fetch-extras`](https://github.com/sindresorhus/fetch-extras).

> FWIW this person on x implemented a tiny wrapper which uses fetch if the client supports upload progress over fetch, with fallback to XMLHTTPRequest

Yep, that's a good approach.

> I'm not sure if ky could do something like that, e.g. whether ky's API surface can be satisfied using an XMLHttpRequest backend (in addition to fetch).

It's doable, especially if the goal is only to support this feature and not be a general purpose fallback for fetch. However, there are environments that only support fetch and not XMLHttpRequest, such as Deno, Node, and Cloudflare Workers. So it's only a partial solution.

Whether the complexity is worth it, I'm not sure. I would be happy to look at a PR.

**Comment by nichoth** [comment]
>  Ky exporting its feature detection mechanisms

Would be a big time saver for me, at least.

**Comment by sholladay** [comment]
Good news! #632 has been merged, so this will be coming in the next release. 🙂
